library/posture.c

```diff
@@ -545,3 +585,273 @@ static void ziti_pr_handle_process(ziti_context ztx, char *id, char *path, bool
+    uv_free_interface_addresses(info, count);
+}
+
+
+
+static void default_pq_domain(ziti_context ztx, const char* id, ziti_pr_domain_cb cb) {
+#if _WIN32
+    uint32_t status;
+    LPWSTR buf;
+    NetGetJoinInformation(NULL, &buf, &status);
+    char domain[256];
+    sprintf_s(domain, sizeof(domain), "%ls", buf);
+    cb(ztx, id, domain);
+    NetApiBufferFree(buf);
+#else
+    cb(ztx, id, "");
+#endif
+}
+
+static void default_pq_process(ziti_context ztx, const char *id, const char *path, ziti_pr_process_cb cb) {
+    bool is_running = false;
+
+    unsigned char *digest;
+    size_t digest_len;
+    char *sha512_hash = NULL;
+    char **signers = NULL;
+    int signers_count = 0;
+
+    uv_fs_t file;
+    int rc = uv_fs_stat(ztx->loop, &file, path, NULL);
+    if (rc != 0) {
+        cb(ztx, id, path, is_running, NULL, NULL, 0);
+        return;
+    }
+
+    is_running = check_running(ztx->loop, path);
+
+    if (hash_sha512(ztx->loop, path, &digest, &digest_len) == 0) {
+        hexify(digest, digest_len, 0, &sha512_hash);
+        ZITI_LOG(VERBOSE, "file(%s) hash = %s", path, sha512_hash);
+    }
+    signers = get_signers(path, &signers_count);
+
+    cb(ztx, id, path, is_running, sha512_hash, signers, signers_count);
+
+    if (sha512_hash) free(sha512_hash);
+    if (signers) {
+        for (int i = 0; i < signers_count; i++) {
+            free(signers[i]);
+        }
+        free(signers);
+    }
+}
+
+static int hash_sha512(uv_loop_t *loop, const char *path, unsigned char **out_buf, size_t *out_len) {
+    size_t digest_size = crypto_hash_sha512_bytes();
+    unsigned char *digest = NULL;
+    int rc = 0;
+
+#define CHECK(op) do{ rc = (op); if (rc != 0) { \
+ZITI_LOG(ERROR, "failed hashing op[" #op "]: %d", rc); \
+goto cleanup;                                   \
+} }while(0)
+
+    uv_fs_t ft;
+    uv_file file = uv_fs_open(loop, &ft, path, UV_FS_O_RDONLY, 0, NULL);
+
+    if (file < 0) { return -1; }
+    uv_buf_t buf = uv_buf_init(malloc(64 * 1024), 64 * 1024);
+    int64_t offset = 0;
+    crypto_hash_sha512_state sha512;
+    crypto_hash_sha512_init(&sha512);
+
+    // hash data
+    while (true) {
+        int read = uv_fs_read(loop, &ft, file, &buf, 1, offset, NULL);
+        if (read == 0) {
+            break;
+        }
+
+        if (read < 0) {
+            rc = -1;
+            goto cleanup;
+        }
+
+        offset += read;
+        CHECK(crypto_hash_sha512_update(&sha512, (uint8_t *) buf.base, read));
+    }
+    digest = malloc(digest_size);
+    CHECK(crypto_hash_sha512_final(&sha512, digest));
+
+    *out_buf = digest;
+    *out_len = digest_size;
+
+    cleanup:
+    if (rc != 0) FREE(digest);
+    uv_fs_close(loop, &ft, file, NULL);
+
+    return rc;
+}
+
+static bool check_running(uv_loop_t *loop, const char *path) {
+    bool result = false;
+#if _WIN32
+    HANDLE sh = CreateToolhelp32Snapshot(TH32CS_SNAPPROCESS, 0);
+    if (sh == NULL) {
+        ZITI_LOG(ERROR, "failed: %d", GetLastError());
```

Comment:
why is the rest of the block allowed to continue if this fails? will line 763 still succeed if if `sh` is NULL?